File: ios/ExpoZendeskLibraryModule.swift
import ExpoModulesCore

import ZendeskSDKMessaging
import ZendeskSDK
// import ZendeskSDKLogger

public class ExpoZendeskLibraryModule: Module {

  // public required init(appContext: AppContext) {
  //   super.init(appContext: appContext)

  //   Logger.enabled = true
  //   Logger.level = .debug
  // }

  public func definition() -> ModuleDefinition {
    Name("ExpoZendeskLibrary")

    AsyncFunction("initialize") { (channelKey: String, promise: Promise) in
      Zendesk.initialize(withChannelKey: channelKey, messagingFactory: DefaultMessagingFactory()) { result in
        switch result {
          case .success:
              promise.resolve("Messaging initialized successfully.")
          case .failure(let error):
              promise.reject("INITIALIZATION_ERROR", error.localizedDescription)
        }
      }
    }

    AsyncFunction("login") { (jwt: String, promise: Promise) in
      Zendesk.instance?.loginUser(with: jwt) { result in
        switch result {
          case .success(let user):
            promise.resolve([
              "id": user.id,
              "externalId": user.externalId
            ])
          case .failure(let error):
            promise.reject("LOGIN_ERROR", error.localizedDescription)
        }
      }
    }

    AsyncFunction("logout") { (promise: Promise) in
      Zendesk.instance?.logoutUser { result in
        switch result {
          case .success(_):
            promise.resolve(nil)
          case .failure(let error):
            promise.reject("LOGOUT_ERROR", error.localizedDescription)
        }
      }
    }

    Function("showConversation") { () -> Void in
      DispatchQueue.main.async {
        guard let messagingViewController = Zendesk.instance?.messaging?.messagingViewController() else {
          print("Failed to get Zendesk messaging view controller.")
          return
        }

        guard let currentViewController = self.appContext?.utilities?.currentViewController() else {
          print("Failed to get the current view controller from appContext.")
          return
        }

        // ✅ Wrap the messaging view in a navigation controller
        let navController = UINavigationController(rootViewController: messagingViewController)
        navController.modalPresentationStyle = .fullScreen

        currentViewController.present(navController, animated: true, completion: nil)
      }
    }
  }
}

File: android/src/main/java/expo/modules/zendesklibrary/ExpoZendeskLibraryModule.kt
package expo.modules.zendesklibrary

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

import zendesk.android.Zendesk;
import zendesk.messaging.android.DefaultMessagingFactory;
import zendesk.messaging.android.push.PushNotifications

class ExpoZendeskLibraryModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoZendeskLibrary")

    Function("showConversation") {    
      Zendesk.instance.messaging.showMessaging(currentActivity)
    }

    Function("setPushNotificaitonToken") { token: String ->
      PushNotifications.updatePushNotificationToken(token)
    }


    AsyncFunction("login") { jwt: String, promise: Promise ->
      Zendesk.instance.loginUser(jwt = jwt,
        successCallback = { user -> 
          promise.resolve(mapOf(
            "id" to user.id,
            "externalId" to user.externalId
          ))
        },
        failureCallback = { error -> 
          promise.reject("LOGIN_ERROR", "Zendesk login failed", error) 
        }
      )
    }

    AsyncFunction("logout") { promise: Promise ->
      Zendesk.instance.logoutUser(
        successCallback = { _ ->
          promise.resolve(null) 
        },
        failureCallback = { error -> 
          promise.reject("LOGOUT_ERROR", "Zendesk logout failed", error) 
        }
      )
    }

    AsyncFunction("initialize")  { channelKey: String, promise: Promise ->
      Zendesk.initialize(
        context = reactContext,
        channelKey,
        successCallback = { _ ->
            promise.resolve(null)
        },
        failureCallback = { error ->
            promise.reject("INITIALIZATION_ERROR", "Zendesk initialization failed", error)
        },
        messagingFactory = DefaultMessagingFactory()
      )
    }


  }

  private val reactContext
    get() = appContext.reactContext ?: throw IllegalStateException("React context is not available")

  private val currentActivity
    get() = appContext.currentActivity ?: throw IllegalStateException("Current activity is not available")
}
